server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuth.scala

```diff
@@ -74,11 +74,40 @@ object DigestAuth {
     challenged(challenge(realm, store, nonceKeeper))
   }
 
+  /** Similar to [[apply]], but exposing the underlying [[challenge]]
+    * [[cats.data.Kleisli]] instead of an entire [[AuthMiddleware]]
+    *
+    * @param realm The realm used for authentication purposes.
+    * @param store A partial function mapping (realm, user) to the
+    *              appropriate password.
+    * @param nonceCleanupInterval Interval (in milliseconds) at which stale
+    *                             nonces should be cleaned up.
+    * @param nonceStaleTime Amount of time (in milliseconds) after which a nonce
+    *                       is considered stale (i.e. not used for authentication
+    *                       purposes anymore).
+    * @param nonceBits The number of random bits a nonce should consist of.
+    */
+  def challenge[F[_], A](
```

Comment:
Probably that was addressed already, but why don't we add a public constructor for `NonceKeeper` instead?

Reply:
That's an option too, although we have to be careful because creating a `NonceKeeper` is a side-effect. I think it depends whether a `NonceKeeper` should ever be shared between auth middlewares or expose any public methods. If not, it's an internal implementation detail.

There's already a signature that works without exposing the `NonceKeeper`, so we can imitate it for `challenge`. We can always expose the `NonceKeeper` in the future if we need to.
https://github.com/http4s/http4s/blob/548d2dd147fb895edcbe19c53af468d5bef590c9/server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuth.scala#L65-L73

Reply:
My thoughts around this were that exposing `NonceKeeper` at all (currently the whole type is package private) incurred more change to the surface area of the package than just tweaking the arguments to a method of `DigestAuth`.

No need to expose complexity that doesn't benefit the users in some way.